### PR TITLE
8342444: Shenandoah: Uncommit regions from a separate, STS aware thread

### PR DESCRIPTION
**Pull request by earthling-amzn**
Currently, Shenandoah uncommits regions from its control thread. The control thread is responsible for starting GC cycles in a timely fashion. Uncommitting memory from this thread may introduce unwanted delays in the control thread's response to GC pressure.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342444](https://bugs.openjdk.org/browse/JDK-8342444): Shenandoah: Uncommit regions from a separate, STS aware thread (**Enhancement** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Author) Review applies to [997360ac](https://git.openjdk.org/jdk/pull/22019/files/997360ac07e5fe5277794ed28517ab5215d7cc69)
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - **Reviewer**) Review applies to [997360ac](https://git.openjdk.org/jdk/pull/22019/files/997360ac07e5fe5277794ed28517ab5215d7cc69)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22019/head:pull/22019` \
`$ git checkout pull/22019`

Update a local copy of the PR: \
`$ git checkout pull/22019` \
`$ git pull https://git.openjdk.org/jdk.git pull/22019/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22019`

View PR using the GUI difftool: \
`$ git pr show -t 22019`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22019.diff">https://git.openjdk.org/jdk/pull/22019.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22019#issuecomment-2471160191)
</details>

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (kdnilsen) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:555]: Was it a bug that previous version of this code did not acquire the heap lock?

Is the lock required for the entirety of time that we are clearing the bitmap?  Or is it just required to get a trustworthy check on is_bitmap_slice_committed()?

Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:555]: After reading more of this PR, I believe we need the heap lock to get a reliable signal of bitmap_slice_committed().  But I believe we do not need the heap lock for ctx->clear_bitmap(region) so would prefer to move that outside the lock, unless I am misunderstanding.

Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:555]: Hmm, I'm not sure we can do that. Prior to this change, the control thread performed both clearing the bitmap and uncommitting the region's bitmap, so they could never happen concurrently. With this change, a separate thread could perform the uncommit. Consider:

1. Control thread takes heap lock, observes that bitmap slice for region A is committed
2. Control thread releases heap lock, begins clearing bitmap (writing zeros to bitmap slice)
3. Uncommit thread takes heap lock, believes it must uncommit region A
4. Uncommit thread uncommits bitmap slice for region A
5. Segfault in Control Thread

I do _believe_ if we had a per region lock, it would be useful here. Holding a lock over the entire heap for this feels like overkill. Or, we could schedule the uncommit so that it does not occur during a GC cycle.

**Comment by earthling-amzn** [comment]
I modified the testing pipelines to set `-Xms4g -Xmx10g -XX:+ShenandoahUncommit`. All performance and stress tests completed successfully on x86 and aarch64. Marking this as ready for review.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:555]: So, wait a sec. This code is in `ShenandoahResetBitmapTask`, so it can run in parallel. Putting a lock here inhibits parallelism. I understand the failure mode, but I think we should really be optimizing for the case when `ShenandoahUncommit` is not enabled (e.g. `-Xmx` == `-Xms`).

Sounds like there is a hassle in allowing concurrent uncommit to overlap with the GC cycle. In addition to this particular problem, we might be stealing cycles from the GC threads and take additional TTSP lag to park the uncommitter for the in-cycle GC pauses. I have no clear solution for this yet, but I think we need to explore if we can suspend the uncommit before going into GC cycle...

File: src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp:55]:
```suggestion
      double shrink_before = (soft_max_changed || explicit_gc_requested) ? 
         current : 
         current - ((double) ShenandoahUncommitDelay / 1000.0);
```

File: src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp:69]:
```suggestion
        locker.wait((int64_t)shrink_period);
```

File: src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.hpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.hpp:37]: Which one of these can be `const`?

**Review by shipilev — COMMENTED** [review]
Cursory review:

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:556]: We could have the control and uncommit threads coordinate their efforts. In the worst case, it could mean delaying concurrent reset while the control thread waits for the uncommit thread to yield.

We could also try a more targeted lock only for the region's bitmap slice, but it doesn't seem right that one thread would be trying to clear a bitmap, while the other is trying to uncommit it. A lock could preserve technical correctness, but contention here would just mean that one thread would have wasted its time (either clearing a bitmap that is then uncommitted, or attempting to clear a bitmap that was first uncommitted (in this case, we would need the control thread to detect this and skip the region)).

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1934]: I'm not sure the order matters here. `ConcurrentGCThread::stop` waits until the target thread sets `_has_terminated`.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp]: Can we open a ticket to consider future improved concurrency by moving clear_bitmap(region) outside the global heap lock?

**Review by kdnilsen — APPROVED** [review]
I'm ok with this, but best to wait for @shipilev approval before integrating.

**Review by kdnilsen — APPROVED** [review]
Have read through the latest version of the code.  Thanks.

**Review by ysramakrishna — APPROVED** [review]
Looks good to me. A few documentation comment requests.

Also please share performance data in this PR or in the ticket, especially from the perf/benchmark that may have precipitated this change.

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:253]: This looks to be happening on every iteration, even if no GC happened. Should this `allow_uncommit()` go into the same block where `allow_commit()` is? Maybe it would be cleaner to make a `StackObj` mark object to manage this state -- up to you.

File: src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp:189]: New line at the end of file here.

**Review by shipilev — APPROVED** [review]
I like it, thanks!

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:253]: Hmm, good catch. I think it's a little worse than this even. The code that is meant to trigger an uncommit:
```C
    if (ShenandoahUncommit) {
      if (heap->check_soft_max_changed()) {
        heap->notify_soft_max_changed();
      } else if (is_gc_requested) {
        heap->notify_explicit_gc_requested();
      }
    }
```
Will only happen when uncommit is forbidden. The notification will fall on deaf ears, as it were. I'll fix this.

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:45]: This is `class`, not `struct`, right? A common name for these in Hotspot are `*No*Mark`, so `ShenandoahNoUncommitMark`?

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1932]: OK, nevermind. We can fix it later if it becomes a problem.

File: src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp:171]: Yeah, at least restore the log format and add `gc+start` log as well.

**Review by shipilev — APPROVED** [review]
I'll approve again, with the following nits:

File: src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp:172]: This one is still not addressed, unfortunately ^

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:267]: Nit: No need for this newline, the sleep logically relates to this block.

**Review by shipilev — APPROVED** [review]
Almost there, modulo restoring the logging.

File: src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp:172]: Yes, I spent some time trying to resurrect `ShenandoahConcurrentPhase` for uncommit here, but it really doesn't want to be used outside of a gc cycle. Also, previously it was logging heap _usage_, which isn't quite what we want here (this may actually increase during this phase, which makes it seem as though nothing is being uncommitted).

I've restored the original logging format, but instead of logging heap usage it is now logging heap committed before and after. Here is an excerpt from specjbb2015 with `-Xms5g -Xmx10g`:

```
[2024-11-20T20:02:25.056+0000][97.396s][22293][info][gc,start       ] Concurrent uncommit
[2024-11-20T20:02:25.072+0000][97.412s][22293][info][gc             ] Concurrent uncommit 5424M->5120M(5120M) 15.988ms
[2024-11-20T20:05:17.916+0000][270.255s][22293][info][gc,start       ] Concurrent uncommit
[2024-11-20T20:05:18.169+0000][270.508s][22293][info][gc             ] Concurrent uncommit 10240M->5120M(5120M) 253.048ms
[2024-11-20T20:06:45.329+0000][357.668s][22293][info][gc,start       ] Concurrent uncommit
[2024-11-20T20:06:45.596+0000][357.935s][22293][info][gc             ] Concurrent uncommit 10240M->5120M(5120M) 267.144ms
[2024-11-20T20:06:57.147+0000][369.486s][22293][info][gc,start       ] Concurrent uncommit
[2024-11-20T20:06:57.148+0000][369.487s][22293][info][gc             ] Concurrent uncommit 5456M->5440M(5440M) 1.189ms
```

Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp]: If we are emitting a log line that looks like a properly formatted GC log line, but the numbers there mean something else for `Concurrent uncommit`, we are bound to confuse users and automatic tools. Uncommit should affect `capacity`, this is how we know how deep we have uncommitted. So, I suggest we emit:

```
Concurrent uncommit XXXXM->XXXXM (YYYYM) z.zzzms
```

...where `XXXX` is the heap used at the end of uncommit (note before and after are the same) and YYYY is capacity. This will not expose users to thinking uncommit grows the heap usage, and would give us instantaneous view on heap usage and capacity.

**Review by shipilev — CHANGES_REQUESTED** [review]
I think log message is still confusing a bit...

File: src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp]: Hmm, the numbers are preceded by `Concurrent uncommit`, with that context it's not much of a stretch to think these numbers represent the change in _committed_ memory. The original log message (in which heap usage may increase during uncommit) was _not_ helpful. A message with the same format in which heap usage also appears to _not change at all_ during an uncommit is also perplexing. Are we trying too hard to preserve the original, not useful message? Maybe we just want a new message that plainly says: 

```
Concurrently uncommitted XXXXM in z.zzzms
```
or
```
Concurrent uncommit: time z.zzzms, committed before XXXXM, committed after YYYYM, capacity ZZZZM
```

File: src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp:75]: I tried to test this on some of my toy examples, and realized this particular line may end up as `locker.wait(0)`, which means "wait indefinitely, until notified". This breaks periodic commits. The old code rode on control thread doing `MAX2(1, ...)`, so we never feed `0` into `wait`. I am also confused about units. The comment above says `shrink_period` is in seconds, but `locker.wait` accepts milliseconds?

Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp:75]: It sounds like this line should be:

```
        locker.wait(MAX2<int64_t>(1, shrink_period * 1000));
```

File: src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp:181]: I think we want an additional space. I see the current output is:

```
[11.366s][info][gc] Concurrent uncommit 32768K(192M) 2.506ms
```

Should probably be:

```
[11.366s][info][gc] Concurrent uncommit 32768K (192M) 2.506ms
```

**Review by shipilev — CHANGES_REQUESTED** [review]

**Comment by earthling-amzn** [comment]
@ysramakrishna - I ran several iterations of specjbb2015 with different variations of polling interval. Results show that 1/10th of `ShenandoahUncommitDelay` is reasonable, and avoids unintentional commit delays when the polling interval is equal or greater than `ShenandoahUncommitDelay`.

```
                      Category |  Count |         Total |      GeoMean |      Average |     Trim 0.1 |       StdDev |      Minimum |      Maximum
openjdk:master   critical_jops |      5 |     50862.000 |    10162.095 |    10172.400 |    10172.400 |      513.605 |     9630.000 |    10882.000                                
30ms polling     critical_jops |      5 |     48035.000 |     9582.113 |     9607.000 |     9607.000 |      778.036 |     8808.000 |    10692.000
30s polling      critical_jops |      5 |     56398.000 |    11272.026 |    11279.600 |    11279.600 |      460.355 |    10627.000 |    11842.000
no polling       critical_jops |      5 |     55917.000 |    11176.046 |    11183.400 |    11183.400 |      460.960 |    10899.000 |    11995.000
```

**Review by shipilev — APPROVED** [review]

File: src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp]: Sorry, I missed your comments here. I noticed the same and have refactored this code.

**Pull request closed without merging**
